### PR TITLE
Moved \ArrayIterator to \Iterator type hints

### PR DESCRIPTION
**Pull request by gplanchat**
**Description (for Contributor and Core Developer)**

I changed some return type hints of connector methods using `\ArrayIterator`, I instead replaced them with `\Iterator`.

Now, you can use `\Generator`s... or whatever iterator you desire.

**Definition Of Done (for Core Developer only)**

| Q | A |
| --- | --- |
| Added Specs |  |
| Added Behats |  |
| Changelog updated |  |
| Review and 2 GTM |  |
| Micro Demo to the PO (Story only) |  |
| Migration script |  |
|  Tech Doc |  |

### CODE REVIEW

File: src/Oro/Bundle/SecurityBundle/Acl/Persistence/AclPrivilegeRepository.php [hunk 1]
```diff
@@ -445,7 +445,7 @@ protected function findAcls(SID $sid, array $oids)
      */
     protected function sortPrivileges(ArrayCollection &$privileges)
     {
-        /** @var \ArrayIterator $iterator */
+        /** @var \Iterator $iterator */
```
Comment (phaseinducer) [inline src/Oro/Bundle/SecurityBundle/Acl/Persistence/AclPrivilegeRepository.php]: could be dropped

Comment (jmleroux) [inline src/Oro/Bundle/SecurityBundle/Acl/Persistence/AclPrivilegeRepository.php]: should be dropped ;)

**Comment by nidup** [comment]
Hi @gplanchat 

Ok to merge this improvement :+1:

However, 4 scenario are broken in CE,
- Add products to variant group via a form.Add products to a variant group
- Apply restrictions when mass editing products with variant groups.Add products to a variant group
- Apply restrictions when mass editing products with variant groups.Add products to a variant group with invalid axis
- Apply restrictions when mass editing products with variant groups.Add products to a variant group with duplicated variant axis values in selection (and not yet in variant group)

The 4 are related to the following exception,
```
  [Symfony\Component\Debug\Exception\FatalThrowableError]                                                                                                                 
  Type error: Argument 1 passed to Pim\Bundle\EnrichBundle\Connector\Reader\MassEdit\FilteredVariantGroupProductReader::detachProducts() must be of the type array, object given, called in /home/nico/git/pim-ce-17-orm/src/Pim/Bundle/EnrichBundle/Connector/Reader/MassEdit/FilteredVariantGroupProductReader.php on line 337
```

Could you rebase and fix the PR please?

PS: we're re-working our CI to be able to open it to contributors ;)

**Comment by aRn0D** [comment]
Good job Grégory @gplanchat. There are some fails on the CI. Could you fix them, please?

**Comment by gplanchat** [comment]
What are the issues on the CI ?

Travis is green on 5.6 and 7.0. 7.1 fails for BC break, requiring a fix I pushed in another PR (#5070, see https://github.com/akeneo/pim-community-dev/pull/5070#issuecomment-276625617).

**Comment by aRn0D** [comment]
I spoke about fails in Nicolas's comment but I will re-run the CI!

**Comment by aRn0D** [comment]
@gplanchat  here the failures : 

* [ee-odm] Successfully update a read only attribute through rule execution - Display available field options
* [ee-odm] Add products to a variant group - Add products to variant group via a form
* [ee-odm] Successfully apply rules after a mas edit operation only on edited products - Apply rules after a mass edit have run
* [failed-to-read] - TEST-mass_action-mass_edit_variant_group_products_with_rules_36.xml
* [ee-odm] Add products to a variant group with duplicated variant axis values in selection (and not yet in variant group) - Apply restrictions when mass editing products with variant groups
* [ee-odm] Successfully execute a rule with copier actions to update non empty values on reference data attributes - Update a single product by applying rules
* [ee-odm] Successfully unpublish several products with reference data - Unpublish many products at once
* [ee-odm] Successfully execute a rule with a setter action and a NOT EMPTY condition on multiple products - Update multiple product by applying rules
* [ee-odm] Successfully execute a rule with a "starts with" condition - Read a single product by applying rules
* [ee-odm] Successfully execute a rule with an "equals" condition - Read a single product by applying rules
* [ee-odm] Successfully execute a rule with an "ends with" condition - Read a single product by applying rules
* [ee-odm] Successfully execute a rule with a "contains" condition - Read a single product by applying rules
* [ee-odm] Successfully execute a rule with a "does not contain" condition - Read a single product by applying rules
* [ee-odm] Successfully execute a rule with setter actions to update non empty values on all kind of fields - Read a single product by applying rules
* [ee-odm] Successfully execute a rule with a "not equal" condition - Read a single product by applying rules
* [ee-odm] Successfully execute a rule with copier actions to update empty values on all kind of attributes - Read a single product by applying rules
* [ee-odm] Successfully execute a rule with a "not empty" condition - Read a single product by applying rules
* [failed-to-read] - TEST-rules-edit_single_product_222.xml
* [ee-odm] Successfully display values coming from rules engine - Ensure variant group values priority when execute a rule
* [ee-odm] Successfully update product and apply rule - Import products with rules
* [failed-to-read] - TEST-rules-import-import_products_with_rules_44.xml
* [ee-odm] Successfully display history after executing a rule - Update product history when rules are executed
* [failed-to-read] - TEST-rules-update_history_when_executing_rules_53.xml
* [ee-odm] Successfully publish all products - Publish many products at once
* [ee-odm] Succesfully mass publish associated product - Publish many products at once
* [ee-odm] Allow to mass publish two products that are associated in two ways (jackadi => unionjack, unionjack => jackadi), I should be able to publish them twice - Publish many products at once
* [ee-odm] Allow to mass publish two products that are associated, I should be able to publish them twice - Publish many products at once
* [ee-odm] Only publish products on which user is the owner - Publish many products at once
* [failed-to-read] - TEST-mass-action-add_products_to_variant_groups_7.xml
* [ce-odm] Add products to a variant group - Apply restrictions when mass editing products with variant groups
* [ce-odm] Add products to a variant group with duplicated variant axis values in selection (and not yet in variant group) - Apply restrictions when mass editing products with variant groups

File: src/Oro/Bundle/SecurityBundle/Acl/Persistence/AclPrivilegeRepository.php (hunk 1 above)
Comment (damien-carcel) [inline src/Oro/Bundle/SecurityBundle/Acl/Persistence/AclPrivilegeRepository.php]: Yep, no PHPDoc inside methods :-) .

**Comment by damien-carcel** [comment]
Hello @gplanchat.

Again, thank you for this PR, it will be great to merged it. Do you want some help to fix the broken behats?

Btw, we merged a lot of stuff on master recently, so you should rebase your branch so we launch a new CI.

Regards,
Damien

**Pull request closed without merging**